File: 5_kyu/Esolang_Interpreters_2_Custom_Smallfuck_Interpreter.go
package kata

// All commands are implemented as functions on the machine structure
type command = func(work *machine)

// Dataset of the running Smallfuck machine
// The code string is stored as array of command functions
type machine struct {
	tape           []bool
	tapePointer    int
	program        []command
	commandPointer int
}

// Command No operation
func commandNop(work *machine) {
	work.commandPointer++
}

// Command > - Move pointer to the right (by 1 cell)
func commandMoveRight(work *machine) {
	work.tapePointer++
	work.commandPointer++
}

// Command < - Move pointer to the left (by 1 cell)
func commandMoveLeft(work *machine) {
	work.tapePointer--
	work.commandPointer++
}

// Command * - Flip the bit at the current cell
func commandFlip(work *machine) {
	work.tape[work.tapePointer] = !work.tape[work.tapePointer]
	work.commandPointer++
}

// Generate command [ - Jump past matching ] if value at current cell is 0
func generateJumpPast(targetCommandPointer int) command {
	return func(work *machine) {
		if !work.tape[work.tapePointer] {
			work.commandPointer = targetCommandPointer
		} else {
			work.commandPointer++
		}
	}
}

// Generate command ] - Jump back to matching [ (if value at current cell is nonzero)
func generateJumpBack(targetCommandPointer int) command {
	return func(work *machine) {
		if work.tape[work.tapePointer] {
			work.commandPointer = targetCommandPointer
		} else {
			work.commandPointer++
		}
	}
}

// Translate tape string
func translateTape(tape string) []bool {
	runes := []rune(tape)
	result := make([]bool, len(runes))
	for i, r := range runes {
		result[i] = r == '1'
	}
	return result
}

// Translate internal []bool to tape string
func translateTapeToString(tape []bool) string {
	result := make([]rune, len(tape))
	for i, t := range tape {
		if t {
			result[i] = '1'
		} else {
			result[i] = '0'
		}
	}
	return string(result)
}

// Translate code string
func translateCode(code string) []command {
	program := make([]command, 0, len(code))
	jumpStack := make([]int, 0)
	for _, c := range code {
		switch c {
		case '>':
			program = append(program, commandMoveRight)
		case '<':
			program = append(program, commandMoveLeft)
		case '*':
			program = append(program, commandFlip)
		case '[':
			jumpStack = append(jumpStack, len(program))
			program = append(program, commandNop)
		case ']':
			i := len(jumpStack) - 1
			jumpStart := jumpStack[i]
			program[jumpStart] = generateJumpPast(len(program))
			program = append(program, generateJumpBack(jumpStart))
			jumpStack = jumpStack[:i]
		}
	}
	return program
}

// run the program
func run(work *machine) {
	nc := len(work.program)
	nt := len(work.tape)
	for work.commandPointer < nc && 0 <= work.tapePointer && work.tapePointer < nt {
		work.program[work.commandPointer](work)
	}
}

func Interpreter(code, tape string) string {
	work := machine{
		tape:           translateTape(tape),
		tapePointer:    0,
		program:        translateCode(code),
		commandPointer: 0}
	run(&work)
	return translateTapeToString(work.tape)
}

File: 5_kyu/Esolang_Interpreters_2_Custom_Smallfuck_Interpreter.lua
-- Convert smallfuck code into lua code.
local function smallfuck2lua(code,tape)
  -- Convert tape
  program = "local tape={ "
  for t in tape:gmatch("[01]") do
    if t == "1" then
      program = program .. "true,"
    else
      program = program .. "false,"
    end
  end
  program = program:sub(1,#program-1) .. "}\n"
  -- Convert code
  program = program.."local i = 1\n"
  for step in code:gmatch(".") do
    if step == ">" then
      program = program.."i=i+1; if i>#tape then return tape end\n"
    elseif step == "<" then
      program = program.."i=i-1; if i<1 then return tape end\n"
    elseif step == "*" then
      program = program.."tape[i]=not tape[i]\n"
    elseif step == "[" then
      program = program.."while tape[i]==true do\n"
    elseif step == "]" then
      program = program.."end\n"
    end
  end
  -- Return tape at end of the program
  program = program .. "return tape\n\n"
  return program
end

-- Smallfuck interpreter.
-- Arguments:
--    code : Smallfuck program to execute.
--    tape : Run program on this tape.
-- Return:
--    Content of the tape after program run.
function interpreter(code, tape)
  -- Translate into lua code
  local program = smallfuck2lua(code,tape)
  -- load the lua code
  program = load(program)
  -- execute the code
  tape = program()
  -- convert boolean tape into string
  local result = ""
  for _,t in ipairs(tape) do
    if t then
      result=result.."1"
    else
      result=result.."0"
    end
  end
  return result
end

return interpreter
